### ISSUE TITLE
Can't build Dune after v1.0.1 on NetBSD

### ISSUE THREAD

**Issue by rneswold** [ocaml/dune#1522]
I haven't been able to build Dune since version v1.0.1. I was using OPAM v1.2.2 and my installation had all sorts of solver issues, so I haven't been focusing on this particular problem. But I upgraded to OPAM 2.0.1 and am trying to rebuild my installation, so I need Dune to install.

If I try to build v1.1.0, or any later version, I get `executable program file not found` when it runs `boot.exe. --release -j 1`. Running OPAM with `-vvv` shows this failure when building v1.4.0:

```
- '/usr/local/opam/4.07.1+BER/bin/ocamllex.opt' -q src/let-syntax/lexer.mll
- '/usr/local/opam/4.07.1+BER/bin/ocamlc.opt' -g -w -40 -o boot.exe unix.cma threads.cma -I +threads boot.ml
+ /usr/local/opam/4.07.1+BER/.opam-switch/build/dune.1.4.0/./boot.exe "--release" "-j" "1" (CWD=/usr/local/opam/4.07.1+BER/.opam-switch/build/dune.1.4.0)
- executable program file not found
[ERROR] The compilation of dune failed at
        "/usr/local/opam/4.07.1+BER/.opam-switch/build/dune.1.4.0/./boot.exe
        --release -j 1".
```

If I run `boot.exe` in that directory, it makes it almost halfway through before failing:

```ShellSession
$ ./boot.exe --release -j 1
Done: 955/2349 (jobs: 1)executable program file not found
$
```

Something between v1.0.1 and v1.1.0 prevents it from building. Since this is a BSD machine, it doesn't have GNU programs installed by default. For instance, I needed to install GNU `sed` to get OPAM v2.0 to build. Maybe `boot.exe` requires a GNU tool? The error message doesn't indicate what executable is missing.

**Comment by rneswold** [comment]
This should have been posted to the dune project. Sorry.

**Comment by diml** [comment]
That's odd. This error doesn't seem to come from Dune. Dune requires nothing else than the OCaml compiler to build itself.

Could you try to pass `--display verbose` to `./boot.exe`?

**Comment by rneswold** [comment]
```
Running[209]: (cd _build/default && /usr/local/opam/4.07.1+BER/bin/ocamlopt.opt -w -40 -g
              -I src/let-syntax/.pp.eobjs -intf-suffix .ml -no-alias-deps
              -o src/let-syntax/.pp.eobjs/pp.cmx -c -impl src/let-syntax/pp.ml)
Running[210]: (cd _build/default && /usr/local/opam/4.07.1+BER/bin/ocamlopt.opt -w -40 -g
              -o src/let-syntax/pp.exe src/let-syntax/.pp.eobjs/lexer.cmx
              src/let-syntax/.pp.eobjs/pp.cmx)
Running[211]: (cd _build/default && src/let-syntax/pp.exe bin/alias.ml) >
              _build/default/bin/alias.pp.ml
executable program file not found
```

There's no error reported for line 210, yet `src/let-syntax/pp.exe` wasn't built.

**Comment by rneswold** [comment]
I take that back:

```
$  find /usr/local/opam/4.07.1+BER -name pp.exe
/usr/local/opam/4.07.1+BER/.opam-switch/build/dune.1.4.0/_build/default/src/let-syntax/pp.exe
/usr/local/opam/4.07.1+BER/.opam-switch/build/dune.1.3.0/_build/default/src/let-syntax/pp.exe
/usr/local/opam/4.07.1+BER/.opam-switch/build/dune.1.2.1/_build/default/src/let-syntax/pp.exe
/usr/local/opam/4.07.1+BER/.opam-switch/build/dune.1.2.0/_build/default/src/let-syntax/pp.exe
/usr/local/opam/4.07.1+BER/.opam-switch/build/dune.1.1.1/_build/default/src/let-syntax/pp.exe
/usr/local/opam/4.07.1+BER/.opam-switch/build/dune.1.1.0/_build/default/src/let-syntax/pp.exe
```

So it has been building them. Just not finding them?

**Comment by rgrinberg** [comment]
The only place where I see an error like this being thrown is in ocaml's [own source](https://github.com/ocaml/ocaml/blob/trunk/runtime/backtrace_byt.c#L325)

The relevant bit replicated here:

```c
  if (caml_cds_file != NULL) {
    exec_name = caml_cds_file;
  } else {
    exec_name = caml_exe_name;
  }

  fd = caml_attempt_open(&exec_name, &trail, 1);
  if (fd < 0){
    caml_fatal_error ("executable program file not found");
    CAMLreturn0;
  }

```

I have no idea how this error can happen, but it would be really useful if OCaml signaled the file that it failed to read.

**Comment by ejgallego** [comment]
Does it happen with a regular [that is to say, non `BER`] switch?

**Comment by rneswold** [comment]
All failures using OPAM 1.2.2 were with non-BER configurations. You'll note above that it's using `ocamlopt.opt`, not `metaocaml.opt` to build `pp.exe`.

**Comment by rneswold** [comment]
Doing a `git bisect`, I tracked the first failure to 5dd218c0. This is the commit that introduces `pp.exe` to the build.

**Comment by rneswold** [comment]
I was able to successfully build on a NetBSD7 machine! The machine on which I want to use it runs NetBSD6. I guess I need to determine the difference in those machines and/or upgrade it. Since NetBSD6 is deprecated in favor of NetBSD7 and 8, I'll close out this issue.

Sorry for the noise.

**Comment by rneswold** [comment]
FWIW: My NetBSD6 system had the `ulimit` for file handles to 128. Raising it to 1024 allowed `dune` to build. The OCaml standard library assumes any error means the file wasn't found but, in this case, I ran out of file handles.

**Comment by diml** [comment]
I added an option to record the max number of opened fds in #1543, we can start monitoring it and try to improve it.

**Comment by rneswold** [comment]
Thanks!

**Comment by diml** [comment]
This should be properly fixed by #1643, now building Dune itself requires very few fds
